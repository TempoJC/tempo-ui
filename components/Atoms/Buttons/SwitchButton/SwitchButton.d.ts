export default SwitchButton;
/**
 * Use this component to show a switch button.
 */
declare const SwitchButton: import('../../../../../node_modules/react').ForwardRefExoticComponent<import('../../../../../node_modules/react').RefAttributes<any>>;
