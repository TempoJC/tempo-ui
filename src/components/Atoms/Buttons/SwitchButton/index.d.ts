import type { MouseEventHandler, TouchEventHandler, Ref, ReactNode } from "react";
import { CommonProps } from "@/types";

export type SwitchButtonProps = CommonProps & {
  autoFocus?: boolean;
  selected?: boolean;
  disabled?: boolean;
  icon: ReactNode;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  onMouseDown?: MouseEventHandler<HTMLButtonElement>;
  onMouseUp?: MouseEventHandler<HTMLButtonElement>;
  onMouseEnter?: MouseEventHandler<HTMLButtonElement>;
  onMouseLeave?: MouseEventHandler<HTMLButtonElement>;
  onTouchStart?: TouchEventHandler<HTMLButtonElement>;
  onTouchEnd?: TouchEventHandler<HTMLButtonElement>;
  ref?: Ref<HTMLButtonElement>;
  tabIndex?: number;
};

declare function SwitchButton(props: SwitchButtonProps): JSX.Element;

export default SwitchButton;
