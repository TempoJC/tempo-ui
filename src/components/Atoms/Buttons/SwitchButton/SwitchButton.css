.tempo-switch-button {
  padding: calc(var(--size1) - 2px);
  color: var(--color-token-light-content-2);
  width: fit-content;
  background: none;
  transition: background 0.3s;
  cursor: pointer;
  box-sizing: border-box;
  border: 1px solid transparent;
}

.tempo-switch-button:focus-visible {
  border: 1px solid var(--color-token-light-semantic-highlight-2);
}

@media (--viewport-xs-m) {
  .tempo-switch-button {
    padding: calc(var(--size1) + 2px);
  }
}

.tempo-switch-button--dark {
  color: var(--color-token-dark-content-2);
}

.tempo-switch-button:hover {
  color: var(--color-token-light-content-hover);
}

.tempo-switch-button--dark.tempo-switch-button:hover {
  color: var(--color-token-dark-content-hover);
}

.tempo-switch-button--selected {
  background-color: var(--color-token-light-background-active);
}

.tempo-switch-button--dark.tempo-switch-button--selected {
  background-color: var(--color-token-dark-background-active);
}

.tempo-switch-button--disabled {
  color: var(--color-token-light-content-disabled);
  cursor: auto;
  pointer-events: none;
}

.tempo-switch-button--dark.tempo-switch-button--disabled {
  color: var(--color-token-dark-content-disabled);
}

.tempo-switch-button--selected.tempo-switch-button--disabled {
  color: var(--color-token-light-content-disabled);
  background-color: var(--color-token-light-background-disabled);
}

.tempo-switch-button--dark.tempo-switch-button--selected.tempo-switch-button--disabled {
  color: var(--color-token-dark-content-disabled);
  background-color: var(--color-token-dark-background-disabled);
}
