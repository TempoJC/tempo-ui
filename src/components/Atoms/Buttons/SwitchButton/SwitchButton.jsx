import { forwardRef, useEffect } from "react";
import cn from "classnames";
import "./SwitchButton.css";
import { useSyncedRef } from "@/hooks";

/**
 * Use this component to show a switch button.
 */
const SwitchButton = forwardRef(
  (
    {
      className,
      onClick,
      onMouseDown,
      onMouseUp,
      onMouseEnter,
      onMouseLeave,
      onTouchStart,
      onTouchEnd,
      selected = false,
      disabled = false,
      tabIndex,
      autoFocus = false,
      icon,
      testId,
      focused = false,
      color = "light",
    },
    ref,
  ) => {
    const buttonRef = useSyncedRef(ref);

    useEffect(() => {
      if (focused) {
        buttonRef.current.focus();
      }
    }, [focused, buttonRef]);

    return (
      <button
        ref={buttonRef}
        data-testid={testId}
        className={cn("tempo-switch-button", className, {
          "tempo-switch-button--dark": color === "dark",
          "tempo-switch-button--selected": selected,
          "tempo-switch-button--disabled": disabled,
        })}
        onClick={onClick}
        onMouseDown={onMouseDown}
        onMouseUp={onMouseUp}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onTouchStart={onTouchStart}
        onTouchEnd={onTouchEnd}
        disabled={disabled}
        tabIndex={tabIndex}
        autoFocus={autoFocus}
      >
        {icon}
      </button>
    );
  },
);

export default SwitchButton;
