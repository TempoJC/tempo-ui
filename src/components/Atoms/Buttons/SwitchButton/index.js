// packages/components-action/switch-button
export { default } from "./SwitchButton.jsx";
