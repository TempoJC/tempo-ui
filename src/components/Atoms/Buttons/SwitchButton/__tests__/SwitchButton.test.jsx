import { userEvent } from "@testing-library/user-event";
import { render } from "@testing-library/react";
import SwitchButton from "../SwitchButton";

const MockIcon = ({ selected, disabled }) => (
  <div data-icon-type={selected ? "selected" : "unselected"} data-icon-state={disabled ? "disabled" : "enabled"} />
);

describe("@tempo-ui/Atoms/Buttons/SwitchButton", () => {
  it("renders", () => {
    const { asFragment } = render(<SwitchButton icon={<MockIcon />} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe("selected", () => {
    it("injects the selected value set to true", () => {
      const { container } = render(<SwitchButton icon={<MockIcon selected />} selected />);

      expect(container.querySelector("[data-icon-type='selected']")).toBeInTheDocument();
    });

    it("injects the selected value set to false", () => {
      const { container } = render(<SwitchButton icon={<MockIcon selected={false} />} selected={false} />);

      expect(container.querySelector("[data-icon-type='unselected']")).toBeInTheDocument();
    });
  });

  describe("disabled", () => {
    it("injects the disabled value set to true", () => {
      const { container } = render(<SwitchButton icon={<MockIcon disabled />} disabled />);

      expect(container.querySelector("[data-icon-state='disabled']")).toBeInTheDocument();
    });

    it("injects the disabled value set to false", () => {
      const { container } = render(<SwitchButton icon={<MockIcon disabled={false} />} disabled={false} />);

      expect(container.querySelector("[data-icon-state='enabled']")).toBeInTheDocument();
    });

    it("calls the event handler", async () => {
      const clickHandler = jest.fn();

      const { container } = render(<SwitchButton testId="switch-button" icon={<MockIcon />} onClick={clickHandler} />);

      const button = container.querySelector("[data-testid='switch-button']");
      await userEvent.click(button);
      expect(clickHandler).toHaveBeenCalled();
    });

    it("does not call the event handler", async () => {
      const clickHandler = jest.fn();

      const { container } = render(
        <SwitchButton testId="switch-button" icon={<MockIcon disabled />} disabled onClick={clickHandler} />,
      );

      const button = container.querySelector("[data-testid='switch-button']");
      await userEvent.click(button);
      expect(clickHandler).not.toHaveBeenCalled();
    });
  });

  it("adds the additional className", () => {
    const { container } = render(<SwitchButton testId="switch-button" className="customClass" icon={<MockIcon />} />);
    expect(container.querySelector("[data-testid='switch-button']")).toHaveClass("customClass");
  });

  describe("testId prop", () => {
    it("renders the SwitchButton with the specified data-testid attribute", () => {
      const { getByTestId } = render(<SwitchButton icon={<MockIcon />} testId="switch-button" />);
      const buttonElement = getByTestId("switch-button");
      expect(buttonElement).toBeInTheDocument();
      expect(buttonElement).toHaveAttribute("data-testid", "switch-button");
    });
  });
});
