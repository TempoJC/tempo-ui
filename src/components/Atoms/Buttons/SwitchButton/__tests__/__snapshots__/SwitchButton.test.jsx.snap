// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/Atoms/Buttons/SwitchButton renders 1`] = `
<DocumentFragment>
  <button
    class="tempo-switch-button"
  >
    <div
      data-icon-state="enabled"
      data-icon-type="unselected"
    />
  </button>
</DocumentFragment>
`;
